build_modules/lib/src/kernel_builder.dart

```diff
@@ -143,68 +144,72 @@ Future<void> _createKernel(
+/// readable kernel file then it, and all of it's dependents will be categorized
+/// as [sourceDeps] which will have all of their [Module.sources].
+Future<void> _findModuleDeps(
     Module root,
-    List<AssetId> transitiveKernelDeps,
-    List<AssetId> transitiveSourceDeps,
+    List<AssetId> kernelDeps,
+    List<AssetId> sourceDeps,
     BuildStep buildStep,
     String outputExtension) async {
-  final toCrawl = Queue.of([root.primarySource]);
-  final nodes = {root.primarySource: _ModuleNode(root, true)};
+  final resolvedModules = await _resolveTransitiveModules(root, buildStep);
 
-  // Helper for any time we encounter a module which is provided via source and
-  // not kernel.
-  //
-  // For each of those we have to ensure that all parents of that module are
-  // provided via source as well.
-  void updateParents(AssetId nodeId) {
-    var node = nodes[nodeId];
-    for (var parent in node.parents) {
-      var parentNode = nodes[parent];
-      if (!parentNode.sourceOnly) {
-        parentNode.sourceOnly = true;
-        parentNode.parents.forEach(updateParents);
-      }
+  final sourceOnly = await _parentsOfMissingKernelFiles(
+      resolvedModules, buildStep, outputExtension);
+
+  for (final module in resolvedModules) {
+    if (sourceOnly.contains(module.primarySource)) {
+      sourceDeps.addAll(module.sources);
+    } else {
+      kernelDeps.add(module.primarySource.changeExtension(outputExtension));
     }
   }
+}
 
-  // Builds the module graph, the `sourceOnly` property of nodes gets
-  // incrementally updated so we need to wait until we are done to actually
-  // provide the deps.
-  while (toCrawl.isNotEmpty) {
-    var moduleId = toCrawl.removeLast();
-    var node = nodes[moduleId];
-    for (final dep in node.module.directDependencies) {
-      var depNode = nodes[dep];
-      if (depNode == null) {
-        toCrawl.add(dep);
-        var depKernelId = dep.changeExtension(outputExtension);
-        var sourceOnly = !await buildStep.canRead(depKernelId);
-        var depModule = Module.fromJson(json.decode(
-                await buildStep.readAsString(
-                    dep.changeExtension(moduleExtension(root.platform))))
-            as Map<String, dynamic>);
-        depNode = _ModuleNode(depModule, sourceOnly);
-        nodes[dep] = depNode;
-      }
-      depNode.parents.add(moduleId);
-      if (depNode.sourceOnly && !node.sourceOnly) {
-        node.sourceOnly = true;
-        updateParents(node.id);
-      }
+/// The transitive dependencies of [root], not including [root] itself.
+Future<List<Module>> _resolveTransitiveModules(
+        Module root, BuildStep buildStep) =>
+    crawlAsync<AssetId, Module>(
+        [root.primarySource],
+        (id) async => Module.fromJson(jsonDecode(await buildStep.readAsString(
+                id.changeExtension(moduleExtension(root.platform))))
+            as Map<String, dynamic>),
+        (id, module) => module.directDependencies).skip(1).toList();
```

Comment:
lets add a comment about the skip here actually (that it is known to be the root)

Reply:
Done